### PR TITLE
Migrate Klaviyo Source Connector to Low Code

### LINKED ISSUE #28867: klaviyo

### PR DESCRIPTION
**Pull request by topefolorunso**
## What
Migrating Source Klaviyo to Low-Code CDK resolves #28867

## How
Created the connector using the connector buider UI, then imported the manifest.yaml and json schema files to the low code cdk.

## Recommended reading order
1. `manifest.yaml`
2. `schemas/*`

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*

No breaking changes

*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

## Pre-merge Actions

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+    required:
+      - api_key
+      - filter
+      - revision
+    properties:
+      api_key:
+        type: string
+        title: API Key
+        airbyte_secret: true
+        order: 0
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:278]: @topefolorunso For this hackathon the request is to move the connector from Python CDK to low-code. You're introducing new parameters in the spec for Klaviyo connector and this is outside of the scope of the contribution. You're more than welcome to introduce this in a next contribution but for now please remove it.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml (hunk 1 above)
Comment (topefolorunso) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:278]: The new parameters are based on the updated API. The API used in the Python CDK is scheduled to be end-of-lifed on Jan 1, 2024 as per https://developers.klaviyo.com/en/v1-2/reference/api-overview

I will remove them as requested and introduce the changes in a new contribution

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:278]: I appreciate the effort @topefolorunso but moving the connector to low code already is a big change. updating the API will introduce a lot more to review and ensure everything is working.

Comment (topefolorunso) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml]: Oh, I understand that. I have made the requested changes now. Kindly review.

**Comment by marcosmarxm** [comment]
@topefolorunso can you share the output of integration tests?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-klaviyo/metadata.yaml [hunk 2]
```diff
@@ -5,24 +5,20 @@ data:
       - klaviyo.com
   connectorSubtype: api
   connectorType: source
-  definitionId: 95e8cffd-b8c4-4039-968e-d32fb4a69bde
-  dockerImageTag: 0.3.2
+  definitionId: 683d1662-6814-446b-a268-5a6a4edbb4da
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-klaviyo/metadata.yaml:8]: Keep the same definition here.

File: airbyte-integrations/connectors/source-klaviyo/metadata.yaml [hunk 3]
```diff
@@ -5,24 +5,20 @@ data:
...
+  dockerImageTag: 0.1.0
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-klaviyo/metadata.yaml:9]: Bump the version to 0.4.0

File: airbyte-integrations/connectors/source-klaviyo/acceptance-test-config.yml [hunk 4]
```diff
@@ -1,38 +1,44 @@
...
-    tests:
-    - config_path: secrets/config.json
-      configured_catalog_path: integration_tests/configured_catalog.json
-      future_state:
-        future_state_path: integration_tests/abnormal_state.json
-  spec:
-    tests:
-    - spec_path: source_klaviyo/spec.json
-connector_image: airbyte/source-klaviyo:dev
-test_strictness_level: high
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-klaviyo/acceptance-test-config.yml:38]: Don't remove this.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

**Comment by topefolorunso** [comment]
> @topefolorunso can you share the output of integration tests? For me a simple read command is not working.

image

**Comment by archangelic** [comment]
@topefolorunso are you able to get the integration tests working?

**Comment by topefolorunso** [comment]
> @topefolorunso are you able to get the integration tests working?

Not entirely, will update soon @archangelic

**Comment by topefolorunso** [comment]
Made some changes but I am now unable to run integration tests, getting the error below

`ImportError: Error importing plugin "connector_acceptance_test.plugin": No module named 'connector_acceptance_test'`

@marcosmarxm @archangelic

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+  base_retriever:
+    type: SimpleRetriever
+    record_selector:
+      $ref: "#/definitions/selector"
+
+  base_stream:
+    type: DeclarativeStream
+    primary_key:
+      - '{{ parameters.primary_key }}'
```
Comment (archangelic) [inline airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:55]: As far as I can tell, this can be removed as you are already declaring primary keys in the streams below that are inheriting this.

**Comment by archangelic** [comment]
I'm running this through tests and will return with a more in-depth review.

File: airbyte-integrations/connectors/source-klaviyo/metadata.yaml [hunk 6]
```diff
@@ -17,10 +17,11 @@ data:
       enabled: true
     oss:
       enabled: true
+  releaseDate: "02-08-2023"
```
Comment (archangelic) [inline airbyte-integrations/connectors/source-klaviyo/metadata.yaml:20]: This needs to be in YYYY-MM-DD format.

**Review by archangelic — CHANGES_REQUESTED** [review]
Can you update the changelog in `docs/integrations/sources/klaviyo.md`?

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml (hunk 5 above)
Comment (archangelic) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:55]: Once you've removed this, you should be able to run the tests and fix any other issues.

**Comment by topefolorunso** [comment]
Changes made @archangelic, kindly review

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+      type: CursorPagination
+      page_size: 100
+      cursor_value: '{{ response.get("links", {}).get("next", {}) }}'
+      stop_condition: '{{ not response.get("links", {}).get("next", {}) }}'
+  
+  base_incremental_sync:
+    type: DatetimeBasedCursor
+    cursor_field: "{{ parameters.cursor_field }}"
+    cursor_datetime_formats:
+      - "{{ parameters.cursor_datetime_formats }}"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:88]:
```
[35m44:[0m [4.56s] {"type": "TRACE", "trace": {"type": "ERROR", "emitted_at": 1693504634501.5, "error": {"message": "Something went wrong in the connector. See the logs for more details.", "internal_message": "No format in ['{{ parameters.cursor_datetime_formats }}'] matching 2021-05-18 01:29:51", "stack_trace": "Traceback (most recent call last):\n  File \"/airbyte/integration_code/main.py\", line 13, in \n    launch(source, sys.argv[1:])\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 200, in launch\n    for message in source_entrypoint.run(parsed_args):\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 114, in run\n    yield from map(AirbyteEntrypoint.airbyte_message_to_string, self.read(source_spec, config, config_catalog, state))\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 156, in read\n    yield from self.source.read(self.logger, config, catalog, state)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/manifest_declarative_source.py\", line 136, in read\n    yield from super().read(logger, config, catalog, state)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 142, in read\n    raise e\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 122, in read\n    yield from self._read_stream(\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 197, in _read_stream\n    for record in record_iterator:\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 328, in _read_full_refresh\n    for record_data_or_message in record_data_or_messages:\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/declarative_stream.py\", line 104, in read_records\n    yield from self.retriever.read_records(stream_slice)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py\", line 305, in read_records\n    most_recent_record_from_slice = self._get_most_recent_record(most_recent_record_from_slice, stream_data, stream_slice)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py\", line 319, in _get_most_recent_record\n    return current_most_recent if self.cursor.is_greater_than_or_equal(current_most_recent, record) else record\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py\", line 283, in is_greater_than_or_equal\n    return self.parse_date(first_cursor_value) >= self.parse_date(second_cursor_value)\n  File \"/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/declarative/incremental/datetime_based_cursor.py\", line 196, in parse_date\n    raise ValueError(f\"No format in {self.cursor_datetime_formats} matching {date}\")\nValueError: No format in ['{{ parameters.cursor_datetime_formats }}'] matching 2021-05-18 01:29:51\n", "failure_type": "system_error"}}}
```

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml (hunk 7 above)
Comment (topefolorunso) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:88]: What stream is this error for? @marcosmarxm

Comment (topefolorunso) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/manifest.yaml:88]: Got it

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Some comments

**Comment by marcosmarxm** [comment]
Hello @topefolorunso the Klaviyo connector has some peculiarities, as usually all GA connector have:
1. The check connection need to have a special handler to remove the API token from an error response ([check here](https://github.com/airbytehq/airbyte/blob/dc269a06ced3e27b81cb5a9c8abe03a3759cb079/airbyte-integrations/connectors/source-klaviyo/source_klaviyo/source.py#L15-L34)). Expose secrets so need to have a solution.
2. There is a custom transformation applied to date time formats ([check here](https://github.com/airbytehq/airbyte/blob/dc269a06ced3e27b81cb5a9c8abe03a3759cb079/airbyte-integrations/connectors/source-klaviyo/source_klaviyo/streams.py#L155-L162)) which is not supported by the low code framework today. This is the error I'm getting when I run your version today. 
3. The streams flow and global exclusion use a ReverseIncremental sync method not supported by the low code framework today ([check here](https://github.com/airbytehq/airbyte/blob/dc269a06ced3e27b81cb5a9c8abe03a3759cb079/airbyte-integrations/connectors/source-klaviyo/source_klaviyo/streams.py#L255C7-L255C40)). This is causing a problem the incremental is not working for both streams.

The low-code allows you to create custom components and probably achieve the full migration. The problem is today we don't have any examples about how to create these custom components. In my opinion the most complex is the reverse incremental sync, which I can help you next week (I'm going to be OOO until Monday). Let me know if you want to try implement 1 and 2. For 1, you need to implement a custom component for `Check` and for 2 you can use `AddFields` as a base implementation to override all date time formats. The problem is the Transformer component in low-code framework doesn't have access to the schema so you need to send the fields manually as an input. I don't see much trouble for that.

The class is located at `airbyte-cdk/sources/declarative/transformations/add_fields.py`

Here is some high level idea about how that should work:
```python
@dataclass
class DateToRFC3339(RecordTransformation):

       def transform(
        self,
        record: Record,
        config: Optional[Config] = None,
        stream_state: Optional[StreamState] = None,
        stream_slice: Optional[StreamSlice] = None,
    ) -> Record:
        kwargs = {"record": record, "stream_state": stream_state, "stream_slice": stream_slice}
        for parsed_field in self._parsed_fields:
            original_value = record.get(parsed_field)
            new_value = pendulum.parse(original_value).to_rfc3339_string()
            record[parsed_field] = new_value
        return record
```

```yaml
    transformations:
      - type: DateToRFC3339
        fields:
          - path: ["created", "updated", ...]
```

This is only a high-level idea @topefolorunso. Let me know what are your thoughts.

**Comment by topefolorunso** [comment]
@marcosmarxm Thank you for the review.

For 1, I think the error handler should not only apply to the check connection but also every time there is an error when trying to read data in each stream. If that makes sense, then can I use a Custom Error handler that does the API token removal from the error response?

P.S I don't think the declarative schema allows custom check stream - https://github.com/airbytehq/airbyte/blob/1f43520ecc1a5f0c4d10a11e84c5965a87b93a2d/airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml#L223-L242

For 2, I think your high level idea should work. I will try to implement it and get back with the changes

**Comment by marcosmarxm** [comment]
> For 1, I think the error handler should not only apply to the check connection but also every time there is an error when trying to read data in each stream. If that makes sense, then can I use a Custom Error handler that does the API token removal from the error response?

This is already happening. The problem is particular to Check connection for Klaviyo where the API return the URL with the API token.

**Comment by marcosmarxm** [comment]
> How do I implement that since custom check is not supported?

I asked the team what is the best path to follow here.

**Comment by marcosmarxm** [comment]
@topefolorunso Hello, unfortunately, the lowcode framework doesn't support a custom Check function today. For that reason, we need to decline your contribution because it can expose some secrets while using the connector. Your contribution will still count as valid for the Hackathon due to all the effort you have added to the migration and improving the knowledge about the current limitations of the lowcode framework the team needs to work on in the future.
